<a href="https://colab.research.google.com/github/Maurice912/DATA_preprocessing_project/blob/main/notebooks/Data_Scraping.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Step 1 – Load and Explore Kaggle Dataset




In [ ]:
import pandas as pd
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
df = pd.read_csv('/content/drive/MyDrive/Data_preprocessing/population_data.csv')

In [ ]:
# Preview data
print("Preview of the data:")
display(df.head())

# Show dataset shape
print(f"\nDataset shape: {df.shape[0]} rows and {df.shape[1]} columns")

# Show dataset info
print("\nDataset info:")
df.info()

# Summary statistics
print("\nSummary statistics:")
display(df.describe())

# Column names
print("\nColumn names:")
print(df.columns.tolist())

Preview of the data:


id Country (or dependency)  Population 2025 Yearly Change  Net Change  \
0   1                   India       1463865525         0.89%    12929734   
1   2                   China       1416096094        −0.23%  −3,225,184   
2   3           United States        347275807         0.54%     1849236   
3   4               Indonesia        285721236         0.79%     2233305   
4   5                Pakistan        255219554         1.57%     3950390   

   Density (P/Km²)  Land Area (Km²) Migrants (net)  Fert. Rate  Median Age  \
0              492          2973190       −495,753        1.94        28.8   
1              151          9388211       −268,126        1.02        40.1   
2               38          9147420        1230663        1.62        38.5   
3              158          1811570        −39,509        2.10        30.4   
4              331           770880     −1,235,336        3.50        20.6   

  Urban Pop % World Share  
0       37.1%      17.78%  
1       67.5%      17.20%  
2       82.8%       4.22%  
3       59.6%       3.47%  
4       34.4%       3.10%


Dataset shape: 233 rows and 12 columns

Dataset info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 233 entries, 0 to 232
Data columns (total 12 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       233 non-null    int64  
 1   Country (or dependency)  233 non-null    object 
 2   Population 2025          233 non-null    int64  
 3   Yearly Change            233 non-null    object 
 4   Net Change               233 non-null    object 
 5   Density (P/Km²)          233 non-null    int64  
 6   Land Area (Km²)          233 non-null    int64  
 7   Migrants (net)           233 non-null    object 
 8   Fert. Rate               233 non-null    float64
 9   Median Age               233 non-null    float64
 10  Urban Pop %              210 non-null    object 
 11  World Share              233 non-null    object 
dtypes: float64(2), int64(4), object(6)
memory usage: 22.0+ KB

Summary statistics:


id  Population 2025  Density (P/Km²)  Land Area (Km²)  \
count  233.000000     2.330000e+02       233.000000     2.330000e+02   
mean   117.000000     3.532080e+07       487.673820     5.583427e+05   
std     67.405489     1.391905e+08      2412.937646     1.694270e+06   
min      1.000000     5.010000e+02         0.000000     0.000000e+00   
25%     59.000000     4.663300e+05        38.000000     2.830000e+03   
50%    117.000000     5.731206e+06        98.000000     8.220000e+04   
75%    175.000000     2.407458e+07       256.000000     4.103400e+05   
max    233.000000     1.463866e+09     25732.000000     1.637687e+07   

       Fert. Rate  Median Age  
count  233.000000  233.000000  
mean     2.306094   31.896996  
std      1.134316    9.860936  
min      0.690000   14.500000  
25%      1.470000   22.900000  
50%      1.940000   32.800000  
75%      2.950000   39.800000  
max      5.940000   57.400000


Column names:
['id', 'Country (or dependency)', 'Population 2025', 'Yearly Change', 'Net Change', 'Density (P/Km²)', 'Land Area (Km²)', 'Migrants (net)', 'Fert. Rate', 'Median Age', 'Urban Pop %', 'World Share']


In [ ]:
# Check for missing values
print("\nMissing values per column:")
print(df.isna().sum())

# Check for duplicates
print(f"\nNumber of duplicate rows: {df.duplicated().sum()}")

# Quick look at data types and non-numeric columns
print("\nNon-numeric columns:")
print(df.select_dtypes(exclude='number').columns.tolist())



Missing values per column:
id                          0
Country (or dependency)     0
Population 2025             0
Yearly Change               0
Net Change                  0
Density (P/Km²)             0
Land Area (Km²)             0
Migrants (net)              0
Fert. Rate                  0
Median Age                  0
Urban Pop %                23
World Share                 0
dtype: int64

Number of duplicate rows: 0

Non-numeric columns:
['Country (or dependency)', 'Yearly Change', 'Net Change', 'Migrants (net)', 'Urban Pop %', 'World Share']


**Exploration Summary**

The dataset contains 233 rows and 12 columns, each representing a country or dependency with projected population and demographic indicators for 2025. All columns are complete except for Urban Pop %, which has 23 missing values. There are no duplicate entries. Key variables include Population 2025, Fertility Rate, Median Age, and Urban Pop %. Several columns contain non-numeric symbols such as “%” or commas, which will need to be cleaned before analysis. Overall, the dataset is well-structured and provides a clear overview of global population trends, forming a strong base for integration with life expectancy and education data in later steps.









# Step 2 Web Scraping

In [ ]:
import requests

url = "https://www.worldometers.info/demographics/life-expectancy/"
headers = {
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
        "AppleWebKit/537.36 (KHTML, like Gecko) "
        "Chrome/124.0.0.0 Safari/537.36"
    )
}

resp = requests.get(url, headers=headers, timeout=30)
print("Status:", resp.status_code, "| HTML length:", len(resp.text))

# keep the HTML for the next step
html = resp.text


Status: 200 | HTML length: 189368


In [ ]:
import pandas as pd

tables = pd.read_html(html)  # use the HTML we fetched in 6.1
print("Number of tables found:", len(tables))
for i, t in enumerate(tables[:3]):  # quick peek at the first 3
    print(f"\n--- Table {i} preview ---")
    display(t.head(5))


Number of tables found: 1

--- Table 0 preview ---


/tmp/ipython-input-4089111155.py:3: FutureWarning: Passing literal html to 'read_html' is deprecated and will be removed in a future version. To read from a literal string, wrap it in a 'StringIO' object.
  tables = pd.read_html(html)  # use the HTML we fetched in 6.1


Rank           Country  Life Expectancy (both sexes)  \
0     1         Hong Kong                         85.77   
1     2             Japan                         85.00   
2     3       South Korea                         84.53   
3     4  French Polynesia                         84.31   
4     5       Switzerland                         84.23   

   Females Life Expectancy  Males Life Expectancy  
0                    88.39                  83.10  
1                    88.03                  81.99  
2                    87.40                  81.44  
3                    86.74                  82.03  
4                    86.06                  82.34

In [ ]:
# Copy the full table so we keep every column
df_life = tables[0].copy()
df_life.columns = [str(c).strip() for c in df_life.columns]

print("Columns in scraped table:")
print(df_life.columns.tolist())

# Preview the first few rows
display(df_life.head())


Columns in scraped table:
['Rank', 'Country', 'Life Expectancy (both sexes)', 'Females Life Expectancy', 'Males Life Expectancy']


Rank           Country  Life Expectancy (both sexes)  \
0     1         Hong Kong                         85.77   
1     2             Japan                         85.00   
2     3       South Korea                         84.53   
3     4  French Polynesia                         84.31   
4     5       Switzerland                         84.23   

   Females Life Expectancy  Males Life Expectancy  
0                    88.39                  83.10  
1                    88.03                  81.99  
2                    87.40                  81.44  
3                    86.74                  82.03  
4                    86.06                  82.34